src/lib/esnext.array.d.ts

```diff
@@ -0,0 +1,51 @@
+interface ReadonlyArray<T> {
+
+    /**
+     * Calls a defined callback function on each element of an array. Then, flattens the result into
+     * a new array.
+     * This is identical to a map followed by a flatten of depth 1.
+     *
+     * @param callback A function that accepts up to three arguments. The flatMap method calls the
+     * callback function one time for each element in the array.
+     * @param thisArg An object to which the this keyword can refer in the callback function. If
+     * thisArg is omitted, undefined is used as the this value.
+     */
+    flatMap<U, This = undefined> (
+        callback: (this: This, value: T, index: number, array: T[]) => U|U[],
+        thisArg?: This
+    ): U[]
+
+    /**
+     * Returns a new array with all sub-array elements concatenated into it recursively up to the
+     * specified depth. If no depth is provided, flatten method defaults to the depth of 1.
+     *
+     * @param depth The maximum recursion depth
+     */
+    flatten<U>(this: ReadonlyArray<U>, depth?: number): U[];
+  }
+
+interface Array<T> {
+
+    /**
+     * Calls a defined callback function on each element of an array. Then, flattens the result into
+     * a new array.
+     * This is identical to a map followed by a flatten of depth 1.
+     *
+     * @param callback A function that accepts up to three arguments. The flatMap method calls the
+     * callback function one time for each element in the array.
+     * @param thisArg An object to which the this keyword can refer in the callback function. If
+     * thisArg is omitted, undefined is used as the this value.
+     */
+    flatMap<U, This = undefined> (
+        callback: (this: This, value: T, index: number, array: T[]) => U|U[],
+        thisArg?: This
+    ): U[]
+
+    /**
+     * Returns a new array with all sub-array elements concatenated into it recursively up to the
+     * specified depth. If no depth is provided, flatten method defaults to the depth of 1.
+     *
+     * @param depth The maximum recursion depth
+     */
+    flatten<U>(this: U[], depth?: number): U[];

```

Comment:
There is a reason why I wrote some of those overloads :)

With this definition, the result of `[[1, 2], [3, 4]].flatten(2)` is of type `number[][]` instead of `number[]`.

Reply:
@Kovensky thank you for the review! Indeed, I see the purpose of these overloads 😃. I have however two remarks: 
1. In [the overloads you have defined](https://github.com/tc39/proposal-flatMap/blob/master/polyfill.d.ts), the "depth precision" stops at 7. Is this enough? Or too much ? I don't know what the guidelines are.
1.  With the definitions below
```typescript
  flatten<U>(this: U[][][][][][][][], depth: 7): U[];
  flatten<U>(this: U[][][][][][][], depth: 6): U[];
  flatten<U>(this: U[][][][][][], depth: 5): U[];
  flatten<U>(this: U[][][][][], depth: 4): U[];
  flatten<U>(this: U[][][][], depth: 3): U[];
  flatten<U>(this: U[][][], depth: 2): U[];
  flatten<U>(this: U[][], depth?: 1): U[];
  flatten<U>(this: U[], depth: 0): U[];
  flatten<U>(depth?: number): U[];
```
if we have the following scenario: 
```typescript
const depthFourArray = [[[[4]]]];
const flatted = depthFourArray.flatten(4);
```
flatted is of type  {}[] instead of number[] (because the last definition will be used). So, the developer will have to use type assertion.

Reply:
Interesting... I would have expected the `depthFourArray.flatten(4)` to be flagged as the `this` of `.flatten(4)` is not a `U[][][][][]`.

As for stopping at 7, it is an arbitrary stopping point, but I would highly question any use of `.flatten` that needs any depth higher than `2`, _maybe_ `3`.

Reply:
Back to working on this PR
I am going to add the overlaods. Thx!